*: add support for cgroup namespace

cyphar opened the pull request:
The cgroup namespace is a new kernel feature available in 4.6+ that
allows a container to isolate its cgroup hierarchy. This currently only
allows you to hide information from /proc/self/cgroup. But I'm currently
working with upstream on expanding it so that you can modify your
hierarchy inside a user namespace (even a rootless container).

Signed-off-by: Aleksa Sarai <asarai@suse.de>

vbatts commented:
seems fine. Is this needing to wait on upstream linux or runc?

cyphar commented:
This has been merged into Linux and will be released in `4.6`. I'd still recommend waiting on the release. The great thing about this API is that its compile-time switch is `CONFIG_CGROUPS`, so if you have a `>=4.6` kernel with cgroups you automatically get this namespace.

vbatts commented:
oh nice. That smooth the transition for sure.

On Mon, Apr 25, 2016 at 11:35 AM, Aleksa Sarai <notifications@github.com>
wrote:

> This has been merged into Linux and will be released in 4.6. I'd still
> recommend waiting on the release. The great thing about this API is that
> its compile-time switch is CONFIG_CGROUPS, so if you have a >=4.6 kernel
> with cgroups you automatically get this namespace.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/opencontainers/runtime-spec/pull/397#issuecomment-214398495>
>

mrunalp commented:
:+1 We can merge this and add it to runc once released.

wking commented:
On Mon, Apr 25, 2016 at 08:14:24AM -0700, Aleksa Sarai wrote:
>     M schema/defs-linux.json (3)
>     M specs-go/config.go (2)

We also want docs in config-linux.md#namespaces, and possibly an
addition to the example in that section and the full example in
config.md.

cyphar commented:
@wking I've added examples and explanations in the docs.

wking commented on config-linux.md:
```diff
@@ -33,6 +33,7 @@ The following parameters can be specified to setup namespaces:
     * **`ipc`** processes inside the container will only be able to communicate to other processes inside the same container via system level IPC
     * **`uts`** the container will be able to have its own hostname and domain name
     * **`user`** the container will be able to remap user and group IDs from the host to local users and groups within the container
+    * **`cgroup`** the container will have an isolated cgroup hierarchy that it can manage
```
Reference 4.6 as the release that landed this feature?

cyphar replied:
Done.

vishh commented:
How do we handle incompatibility with new kernel features as of now? Is the Spec attempting to provide some means of discovery of the features that a given kernel supports?

mrunalp commented:
@vishh I think probably leave that to the runtime.

vishh commented:
@mrunalp: That would affect client portability right? Is cross-runtime
portability a goal?

On Mon, Apr 25, 2016 at 2:30 PM, Mrunal Patel <notifications@github.com>
wrote:

> @vishh <https://github.com/vishh> I think probably leave that to the
> runtime.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/opencontainers/runtime-spec/pull/397#issuecomment-214529571>
>

cyphar commented:
@vishh We can make it clear in the spec that a runtime implementation must fail if it can't create all of the requested namespaces. Anything outside of that should be left to the runtime to handle (such as figuring out whether or not it can create the namespaces).

wking commented:
9926619 looks good to me.  I don't mind whether the
fail-if-not-supported language [1] lands via this PR or a separate PR.

[1]: https://github.com/opencontainers/runtime-spec/pull/397#issuecomment-214556443

crosbymichael commented:
LGTM

mrunalp commented on config-linux.md:
```diff
@@ -27,12 +27,13 @@ Namespaces are specified as an array of entries inside the `namespaces` root fie
+    * **`ipc`** processes inside the container will only be able to communicate to other processes inside the same container via system level IPC. Support for this was added in Linux 2.6.19.
+    * **`uts`** the container will be able to have its own hostname and domain name. Support for this was added in Linux 2.6.19.
+    * **`user`** the container will be able to remap user and group IDs from the host to local users and groups within the container. Support for this was added in Linux 3.8.
+    * **`cgroup`** the container will have an isolated cgroup hierarchy that it can manage. Support for this was added in Linux 4.6.
```
Can we modify the text here - that it sees an isolated/virtualized view of the cgroup heirarchy? The manage portion doesn't quite work well yet ;)

We can use text from here https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=d4021f6cd41f03017f831b3d40b0067bed54893d

cyphar replied:
Heh, it was a bit ambitious for a description. ;)

cyphar commented:
@mrunalp Updated. I've also added the "must fail" wording as @wking requested.

mrunalp commented on config-linux.md:
```diff
@@ -27,18 +27,22 @@ Namespaces are specified as an array of entries inside the `namespaces` root fie
 Also, when a path is specified, a runtime MUST assume that the setup for that particular namespace has already been done and error out if the config specifies anything else related to that namespace.
 
+If the runtime is unable to create or join all of the requested namespaces, it
+MUST fail so as not to lull the user into a false sense of security.
```
I think we can drop 'so as not to lull the user into a false sense of security'. Otherwise looks fine.

cyphar replied:
Fix'd.

mrunalp commented:
LGTM

wking commented on config-linux.md:
```diff
@@ -27,18 +27,22 @@ Namespaces are specified as an array of entries inside the `namespaces` root fie
 Also, when a path is specified, a runtime MUST assume that the setup for that particular namespace has already been done and error out if the config specifies anything else related to that namespace.
 
+If the runtime is unable to create or join ALL of the requested namespaces, it
+MUST fail.
```
For `consistency with `runtime.md`][1], we should probably use “generate an error”.  Although @vishh's concern was [broader than namespaces][2], so I think we might want to put this language in step 2 of the [lifecycle][3].

Also, [one line per sentence][4] ;).

[1]: https://github.com/opencontainers/runtime-spec/blob/v0.5.0/runtime.md#errors
[2]: https://github.com/opencontainers/runtime-spec/pull/397#issuecomment-214505281
[3]: https://github.com/opencontainers/runtime-spec/blob/v0.5.0/runtime.md#lifecycle
[4]: https://github.com/opencontainers/runtime-spec/blob/v0.5.0/style.md#one-sentence-per-line

cyphar replied:
> Also, one line per sentence ;).

Eww. 80 characters is ~~objectively~~ the right wrapping. ;)
But yes, I'll update the lifecycle documentation then.

cyphar commented:
@wking @mrunalp I've updated things as requested. The documentation is now accurate about what the cgroup namespace does -- and I've added an extra line to the lifecycle documentation to make clear that a runtime must generate an error if it can't set up the environment requested by a container.

wking commented:
On Fri, May 27, 2016 at 05:02:18AM -0700, Aleksa Sarai wrote:
> @wking @mrunalp I've updated things as requested.

Heh, f0e4d33 → a021e63 becomes much more excited about isolation (from
“only really hides” to “invaluable for hiding” ;).

Anyhow, the tip commit (058860b5) looks good (and can be cherry-picked
into a new PR and landed now as far in my view, although I'm not a
maintainer).

I'm not sure about the penultimate commit; will leave an inline comment.

wking commented on config-linux.md:
```diff
@@ -27,17 +27,17 @@ Namespaces are specified as an array of entries inside the `namespaces` root fie
 * **`path`** *(string, optional)* - path to namespace file in the [runtime mount namespace](glossary.md#runtime-namespace)
 
 If a path is specified, that particular file is used to join that type of namespace.
-Also, when a path is specified, a runtime MUST assume that the setup for that particular namespace has already been done and error out if the config specifies anything else related to that namespace.
```
This is new with your [recent reroll][1], and rolls back #158.  What has changed since #158?  The motivation for this line was “[we may support it if someone gives a convincing use-case][2]”, but you're removing it (in a drive-by change?) while adding a `cgroup` namespace, and I don't see the connection.

[1]: https://github.com/opencontainers/runtime-spec/pull/397#issuecomment-222129208
[2]: https://github.com/opencontainers/runtime-spec/pull/158#issuecomment-138997548

cyphar replied:
Ugh, looks like it was a dodgy rebase. I'll fix this up.

cyphar commented:
@wking Fixed the dodgy rebase and I've changed the tone in my commit to be less excited (I forgot to remove the enthusiasm after I realised that the cgroup namespace doesn't allow subtree delegation).

wking commented on runtime.md:
```diff
@@ -35,6 +35,7 @@ The lifecycle describes the timeline of events that happen from when a container
 1. OCI compliant runtime's `create` command is invoked with a reference to the location of the bundle and a unique identifier.
    How these references are passed to the runtime is an implementation detail.
 2. The container's runtime environment MUST be created according to the configuration in [`config.json`](config.md).
+   If the runtime is unable to create the runtimme environment specified in the [`config.json`](config.md) it MUST generate an error.
```
“runtimme” → “runtime”.

wking replied:
Or maybe just drop that second “runtime” instances and say “… the environment specified…”.

wking replied:
Also, you probably want a comma after `config.json`: “If … `config.json`, it MUST generate an error.”

cyphar replied:
Fixed. And I fixed up the email madness.

wking commented:
Everything through 5d3c351c looks pretty good to me (just a few
copy-edit suggestions for the tip commit).

And you're using two different emails, in case you wanted to adjust
those before this lands.

wking commented:
Everything through 3ce281cb looks good to me.

mrunalp commented on config-linux.md:
```diff
@@ -27,12 +27,13 @@ Namespaces are specified as an array of entries inside the `namespaces` root fie
-    * **`ipc`** processes inside the container will only be able to communicate to other processes inside the same container via system level IPC
-    * **`uts`** the container will be able to have its own hostname and domain name
-    * **`user`** the container will be able to remap user and group IDs from the host to local users and groups within the container
+    * **`pid`** processes inside the container will only be able to see other processes inside the same container. Support for this was added in Linux 2.6.24.
```
One sentence per line. Otherwise looks good.

cyphar replied:
Damn, I swore I fixed this.

philips replied:
I don't know if you can do this with lists in md @mrunalp

tianon commented:
As discussed in the call, my $0.02 on whether to include kernel versions of required features in the spec, I don't think it's really strictly necessary for the spec -- having a separate "cheat sheet" might make sense (mapping features to minimum kernel versions), but if the information is reasonably readily available elsewhere, then I think it's sufficient for the tools to simply error out when unsupported features are used (relying more on the runtime check than fuzzy kernel version matching, which RedHat has a habit of throwing a wrench in :smile:).

crosbymichael commented:
Yes, we are not recording the history of the linux kernel in this spec.

mrunalp commented:
@cyphar could you update the PR to address the comments? We kinda need this for 1.0 :)
Thanks!

crosbymichael commented:
I wouldn't say that this is a blocker for 1.0.  Its adding an option and does not affect schema or anything so it can come at any time.

mrunalp commented:
Yeah, not a hard blocker but will be nice to have since it is so close to being merged.

cyphar commented:
So you want me to remove all of the "This is available since Linux X.Y.Z" sentences?

mrunalp commented:
@cyphar Yes. That's what we discussed and agreed to in the OCI call yesterday.

philips commented:
LGTM

Agreed on removing the kernel versions. It is meaningless given some Linux distros do tons of backporting.

[![Approved with PullApprove](https://img.shields.io/badge/pullapprove-approved-brightgreen.svg)](https://pullapprove.com/opencontainers/runtime-spec/pull-request/397/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=opencontainers/runtime-spec)

cyphar commented:
@opencontainers/runtime-spec-maintainers There, I've removed the kernel versions.

crosbymichael commented:
LGTM

crosbymichael commented:
@mrunalp @philips  you have to LGTM again because the commits were updated

vbatts commented:
LGTM
